### PR TITLE
[Perf] CPU LLVM adstack-cache: skip per-launch bump-writes + ndarray_shapes capture on forward-only handles

### PR DESCRIPTION
**Pull request by duburcqa**
# Skip adstack-cache bump-writes for forward-only CPU LLVM launches

> Three commits restoring `single_franka.py --cpu` with `GS_ENABLE_NDARRAY=1` from 5600 FPS back to ~6500-6700 FPS. Each commit narrows the gate one level finer.

## Why

`bump_writes_for_kernel_llvm` (added in #620) walks every task's `snode_writes` / `arr_writes` / `arr_reads` per launch and bumps a per-snode / per-DeviceAllocation gen counter for each id. `LaunchContextBuilder::set_arg_ndarray_*` mirrors per-arg shapes into `ndarray_shapes` for the adstack-overflow diagnose snapshot. Both are dead work on forward-only programs: nothing reads the gen counters and nothing reaches the diagnose path. CPU pays for it visibly because dispatch is synchronous; GPU launchers hide the same cost behind async dispatch.

## What each commit does

- **C1** - one-way `AdStackCache::has_any_recordings()` flag, flipped true by any `record_*`. Gates the CPU `bump_writes_for_kernel_llvm` overload and the three `set_arg_ndarray_*` `ndarray_shapes` writes.
- **C2** - per-snode refinement: `observed_snode_ids_` + `any_external_read_observed_` populated from every `record_*`. Inside CPU `bump_writes_for_kernel_llvm`, snode bumps are skipped for unobserved ids and the arg loops are skipped wholesale when no `ExternalReadObs` has been recorded. Targets mixed programs where C1's program-wide flag is latched true.
- **C3** - per-devalloc refinement: `observed_devalloc_ptrs_` populated from every `ExternalReadObs.observed_devalloc` and per-task `arg_gens` tuple. Inside the bump arg loops, after the unavoidable `array_ptrs.find` resolves an arg slot to a `DeviceAllocation *`, the bump update is skipped when no cached entry observes that devalloc. Targets the `kernel_1(a)` + `kernel_2(b)` + `kernel_2.grad(b)` workflow where `kernel_1`'s args slot through `kernel_2.grad`'s observed-set check and pay only the find, not the bump.

## Per-handle gate considered, rejected

An earlier draft added a fourth commit that precomputed `kernel_has_adstack` at `register_llvm_kernel` and skipped `bump_writes_for_kernel_llvm` entirely for forward-only handles. That gate is unsafe: between two forward/backward pairs (i.e. between training steps), a regular forward-only kernel is allowed to mutate any field or ndarray, including state a later reverse-mode sizer observes. Skipping the bump there would let a previously recorded sizer entry replay as a false cache hit on the next reverse launch (overflow or silently wrong gradient). The loop-bound mutation limitation that bans such writes only applies within a single fwd/bwd pair (a current design limitation, not an intended guarantee); outside that window the bump is correctness-critical. An expanded comment at the bump call site in `kernel_launcher.cpp` records this rationale so the same gate is not reintroduced.

## Correctness

The flag / set transitions are one-way (false to true on first `record_*` / observation). An id (snode_id or devalloc_ptr) becomes "observed" only inside a `record_*`, which snapshots the current gen counter into `observed_gen`; from that launch onward every kernel write to the id passes the gate and bumps, so a later cache lookup detects the mismatch and invalidates. Stale devalloc pointers (Ndarrays destroyed and addresses recycled) at most cost an extra bump on the new binding; the original cache entry self-invalidates on its next lookup since `data_ptr == observed_devalloc` fails against the new value.

## Tests

```
QD_OFFLINE_CACHE=0 python -m pytest tests/python/test_adstack.py -x -n 8 -k arm64
================= 416 passed, 4 xfailed in 16.68s =================
```

### CODE REVIEW

File: quadrants/runtime/cpu/kernel_launcher.cpp [hunk 1]
```diff
@@ -136,8 +136,12 @@ void KernelLauncher::launch_llvm_kernel(Handle handle, LaunchContextBuilder &ctx
     }
   }
   // Adstack-cache invalidation bump - see `bump_writes_for_kernel_llvm` in `program/adstack_size_expr_eval.{h,cpp}`.
-  bump_writes_for_kernel_llvm(executor->get_program(), &ctx, launcher_ctx.snode_writes_per_task,
-                              launcher_ctx.arr_writes_per_task, launcher_ctx.arr_reads_per_task);
+  // Forward-only handles cannot legally invalidate any cached sizer output under the autodiff contract (see
+  // `kernel_has_adstack` doc in `kernel_launcher.h`), so the entire per-launch bump path is skipped for them.
+  if (launcher_ctx.kernel_has_adstack) {
+    bump_writes_for_kernel_llvm(executor->get_program(), &ctx, launcher_ctx.snode_writes_per_task,
+                                launcher_ctx.arr_writes_per_task, launcher_ctx.arr_reads_per_task);
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply quadrants/runtime/cpu/kernel_launcher.cpp]: fixed

**Comment by duburcqa** [comment]
`======== 642 passed, 3 skipped, 2 xfailed in 1117.40s (0:18:37) =======`

**Comment by duburcqa** [comment]
Comparison between [Genesis main + Quadrants v0.7.5] vs [Genesis https://github.com/Genesis-Embodied-AI/genesis-world/pull/2742 + this Quadrants PR]:

|env                     |batch_size|backend|dtype  |gjk_collision|constraint_solver|runtime_fps_main|runtime_fps_685|runtime_fps_delta_pct|
|------------------------|----------|-------|-------|-------------|-----------------|----------------|---------------|---------------------|
|anymal_random           |30000     |cuda   |field  |-            |-                |9302235         |9301503        |-0.01                |
|anymal_uniform          |30000     |cuda   |field  |-            |-                |12090405        |12363794       |+2.26                |
|anymal_uniform_kinematic|0         |cpu    |field  |-            |-                |1837            |1924           |+4.74                |
|anymal_uniform_kinematic|30000     |cuda   |field  |-            |-                |10296266        |10502615       |+2.00                |
|anymal_zero             |0         |cpu    |field  |-            |-                |6501            |6849           |+5.35                |
|anymal_zero             |30000     |cuda   |field  |-            |-                |18869385        |19317580       |+2.38                |
|box_pyramid_3           |4096      |cuda   |field  |-            |-                |974550          |1004902        |+3.11                |
|box_pyramid_4           |4096      |cuda   |field  |-            |-                |390715          |399054         |+2.13                |
|box_pyramid_5           |4096      |cuda   |field  |-            |-                |138068          |141077         |+2.18                |
|box_pyramid_6           |4096      |cuda   |field  |False        |-                |58511           |59743          |+2.11                |
|box_pyramid_6           |4096      |cuda   |field  |True         |-                |60844           |60556          |-0.47                |
|dex_hand                |4096      |cuda   |field  |-            |-                |17177           |17588          |+2.39                |
|duck_in_box_easy        |30000     |cuda   |field  |False        |-                |26378787        |27016185       |+2.42                |
|duck_in_box_easy        |30000     |cuda   |field  |True         |-                |9568199         |9706100        |+1.44                |
|duck_in_box_hard        |0         |cpu    |field  |-            |-                |4947            |5152           |+4.14                |
|duck_in_box_hard        |30000     |cuda   |field  |False        |-                |10138119        |10376434       |+2.35                |
|duck_in_box_hard        |30000     |cuda   |field  |True         |-                |3408111         |3393225        |-0.44                |
|franka                  |30000     |cuda   |field  |-            |-                |20023701        |22251821       |+11.13               |
|franka_accessors        |0         |cpu    |field  |-            |-                |1137            |1196           |+5.19                |
|franka_accessors        |30000     |cuda   |field  |-            |-                |15178267        |16071170       |+5.88                |
|franka_free             |30000     |cuda   |field  |-            |-                |31982335        |33451319       |+4.59                |
|franka_random           |0         |cpu    |field  |-            |-                |5632            |6046           |+7.35                |
|franka_random           |30000     |cuda   |field  |-            |CG               |16433064        |16720194       |+1.75                |
|franka_random           |30000     |cuda   |field  |-            |Newton           |16437955        |16586141       |+0.90                |
|franka_random           |30000     |cuda   |field  |False        |-                |16455936        |16561089       |+0.64                |
|franka_random           |30000     |cuda   |field  |True         |-                |11482652        |11650246       |+1.46                |
|g1_fall                 |4096      |cuda   |field  |-            |Newton           |914720          |921531         |+0.74                |
|go2                     |4096      |cuda   |field  |False        |CG               |3616266         |3684005        |+1.87                |
|go2                     |4096      |cuda   |field  |False        |Newton           |4432073         |4541560        |+2.47                |
|go2                     |4096      |cuda   |field  |True         |-                |3252181         |3361526        |+3.36                |
|shadow_hand_cubes       |0         |cpu    |field  |-            |-                |41              |40             |-2.44                |
|shadow_hand_cubes_sparse|0         |cpu    |field  |-            |-                |66              |66             |+0.00                |
|anymal_random           |30000     |cuda   |ndarray|-            |-                |7701487         |7661304        |-0.52                |
|anymal_uniform          |30000     |cuda   |ndarray|-            |-                |10491936        |10385118       |-1.02                |
|anymal_uniform_kinematic|0         |cpu    |ndarray|-            |-                |1097            |1143           |+4.19                |
|anymal_uniform_kinematic|30000     |cuda   |ndarray|-            |-                |8204543         |8358898        |+1.88                |
|anymal_zero             |0         |cpu    |ndarray|-            |-                |3519            |3637           |+3.35                |
|anymal_zero             |30000     |cuda   |ndarray|-            |-                |14210865        |14203434       |-0.05                |
|box_pyramid_3           |4096      |cuda   |ndarray|-            |-                |760682          |769255         |+1.13                |
|box_pyramid_4           |4096      |cuda   |ndarray|-            |-                |317781          |323092         |+1.67                |
|box_pyramid_5           |4096      |cuda   |ndarray|-            |-                |118110          |118952         |+0.71                |
|box_pyramid_6           |4096      |cuda   |ndarray|False        |-                |52143           |52750          |+1.16                |
|box_pyramid_6           |4096      |cuda   |ndarray|True         |-                |51787           |51862          |+0.14                |
|dex_hand                |4096      |cuda   |ndarray|-            |-                |14111           |14195          |+0.60                |
|duck_in_box_easy        |30000     |cuda   |ndarray|False        |-                |19631916        |19903823       |+1.39                |
|duck_in_box_easy        |30000     |cuda   |ndarray|True         |-                |8492583         |8466776        |-0.30                |
|duck_in_box_hard        |0         |cpu    |ndarray|-            |-                |2960            |3178           |+7.36                |
|duck_in_box_hard        |30000     |cuda   |ndarray|False        |-                |7545036         |7547354        |+0.03                |
|duck_in_box_hard        |30000     |cuda   |ndarray|True         |-                |2797831         |2842460        |+1.60                |
|franka                  |30000     |cuda   |ndarray|-            |-                |15812438        |15915325       |+0.65                |
|franka_accessors        |0         |cpu    |ndarray|-            |-                |926             |900            |-2.81                |
|franka_accessors        |30000     |cuda   |ndarray|-            |-                |11348780        |11577474       |+2.02                |
|franka_free             |30000     |cuda   |ndarray|-            |-                |20975266        |21509160       |+2.55                |
|franka_random           |0         |cpu    |ndarray|-            |-                |3401            |3345           |-1.65                |
|franka_random           |30000     |cuda   |ndarray|-            |CG               |12417711        |12695723       |+2.24                |
|franka_random           |30000     |cuda   |ndarray|-            |Newton           |12463068        |12376790       |-0.69                |
|franka_random           |30000     |cuda   |ndarray|False        |-                |12441446        |12382315       |-0.48                |
|franka_random           |30000     |cuda   |ndarray|True         |-                |8635885         |8719801        |+0.97                |
|g1_fall                 |4096      |cuda   |ndarray|-            |Newton           |713414          |730314         |+2.37                |
|go2                     |4096      |cuda   |ndarray|False        |CG               |2572396         |2639930        |+2.63                |
|go2                     |4096      |cuda   |ndarray|False        |Newton           |3045152         |3065397        |+0.66                |
|go2                     |4096      |cuda   |ndarray|True         |-                |2264723         |2295495        |+1.36                |
|shadow_hand_cubes       |0         |cpu    |ndarray|-            |-                |32              |32             |+0.00                |
|shadow_hand_cubes_sparse|0         |cpu    |ndarray|-            |-                |44              |44             |+0.00                |

[speed_test_main.txt](https://github.com/user-attachments/files/27583557/speed_test_main.txt)
[speed_test_685.txt](https://github.com/user-attachments/files/27583558/speed_test_685.txt)

**Comment by hughperkins** [comment]
[x] file line counts seem ok
[x] Genesis unit tests passing
[x] Genesis benchmarks seem neutral to positive
[x] No user facing doc changes required

=> ok to merge
